### PR TITLE
backupccl: teach online restore to synthesize key prefixes

### PR DESCRIPTION
**Pull request by msbutler**
This patch enables the user to restore into a non-empty cluster by teaching
online restore to rewrite table IDs and rekey all restored keys. As part of
this change, we assert that all restore span entries and backup file span's
contain a common prefix. Because backups and restores of tenants break this
assertion, this patch also disallows online restore's of tenants. A future PR
can re-enable online restores of tenants by updating this rekeying logic.

Fixes https://github.com/cockroachdb/cockroach/pull/119847

Release note (sql change): this patch enables users to run online restore on all kinds of backups created on 24.1 except backups of tenants. 23.2 backups are only online restorable under specific technical conditions, so we do not recommend users run online restore from non 24.1 backups.

### CODE REVIEW

**Comment by msbutler** [comment]
The first commit is being reviewed in https://github.com/cockroachdb/cockroach/pull/119964

File: pkg/ccl/backupccl/restore_planning_test.go [hunk 1]
```diff
@@ -151,24 +153,37 @@ func TestBackupManifestVersionCompatibility(t *testing.T) {
 			minimumSupportedVersion: roachpb.Version{Major: 22, Minor: 2},
 			expectedError:           "the backup is from a version older than our minimum restorable version 22.2",
 		},
+		{
+			name:                    "online-restore-old-backup",
+			backupVersion:           roachpb.Version{Major: 23, Minor: 2},
```
Comment (msbutler) [inline pkg/ccl/backupccl/restore_planning_test.go:158]: @dt Sadly this unit test instead fails with `cannot upgrade to 24.1: node running 23.2`. Does this have something to do with minting 24.1?

**Comment by dt** [comment]
How sure are we that we want to put this hard 24.1 backup requirement here? How much worse would it be to say 24.1 (i.e. prefixless) OR `allRewritesAreNoops` ?

File: pkg/ccl/backupccl/restore_planning.go [hunk 2]
```diff
@@ -1644,7 +1642,14 @@ func checkBackupManifestVersionCompatability(
 	version clusterversion.Handle,
 	mainBackupManifests []backuppb.BackupManifest,
 	unsafeRestoreIncompatibleVersion bool,
+	onlineRestore bool,
 ) error {
+	if onlineRestore && unsafeRestoreIncompatibleVersion {
+		return errors.New("cannot use UNSAFE_RESTORE_INCOMPATIBLE_VERSION with EXPERIMENTAL_DEFERRED_COPY")
+	}
+	if onlineRestore && mainBackupManifests[0].ClusterVersion.Less(clusterversion.V24_1.Version()) {
```
Comment (dt) [inline pkg/ccl/backupccl/restore_planning.go:1650]: Really we should be checking the Elide mode field of the manifest right?

**Comment by msbutler** [comment]
> How sure are we that we want to put this hard 24.1 backup requirement here? How much worse would it be to say 24.1 (i.e. prefixless) OR `allRewritesAreNoops` ?

It seems simpler to only support OR from 24.1 prefixless backups. Supporting `allRewritesAreNoops` feels like a very narrow set of use cases: table and database restores into an empty cluster. I don't see much benefit in supporting this in 24.1 for the cost of extra business logic.

**Comment by dt** [comment]
> Supporting allRewritesAreNoops feels like a very narrow set of use cases: table and database restores into an empty cluster. I don't see much benefit in supporting this in 24.1 for the cost of extra business logic.

a) How much business logic? We already have `allRewritesAreNoops` so if it is true, we just don't need to insist on 24.1, and we just _don't set_ SyntheticPrefix on our requests, right? That doesn't seem _too_ bad to me?
b) Is it _that_ narrow? I think "try this our on your existing backups and see how it works; you'll need to use a fresh cluster to test it, but once you have 24.1 backups you can use existing clusters" is kinda nice? Not so nice I'd keep a ton of complexity/code, but if (a) is that it isn't a lot of code, then it narrow is fine?
c) Selfishly, I just want to be able to use fixtures I'm already using / not blow it up over night. I might be all for deleting this in, say, a month or two, but having 24.1 maintain (limited, i.e. new cluster) flexibility?

**Comment by msbutler** [comment]
my opinion was weakly held as well :D. See the commit "address dt comments". The patch was cleaner than I though it would be -- only required a check in planning.

**Comment by msbutler** [comment]
@dt hrm, I'm changing my mind again about whether we should allow OR from backups with prefixes. I've realized while writing a [test ](https://github.com/cockroachdb/cockroach/blob/e738de1b36625f90a3103d0cee3ba233d6e1274f/pkg/ccl/backupccl/testdata/backup-restore/online-restore-auto-stats) in a separate PR that OR can fail even if the restoring cluster is empty. The table IDs in the backup must match perfectly to the restoring table IDs, which would not occur in this test, because the backed up cluster has 2 dbs and we only restore the 2nd db. This feels like a headache to explain to users.

**Comment by dt** [comment]
I would be fine just _not_ telling users it is possible and then hey, if you're in the case where it works (ids are available) yay. But again, this is a weak preference, so if you want to just say no <24.1 backups ever, I'll be fine with that.

**Comment by msbutler** [comment]
Sounds good, I've squashed the commit that addresses your feedback and added a release note which recommends only running OR using 24.1 backups.

File: pkg/ccl/backupccl/restore_online.go [hunk 3]
```diff
@@ -173,8 +228,8 @@ func sendAddRemoteSSTWorker(
 				// span rather than one we have added to, since we add with estimated
 				// stats and splitting a span with estimated stats is slow.
 				if batchSize+file.BackupFileEntryCounts.DataSize > targetBatchSize {
-					log.Infof(ctx, "flushing %s batch of %d SSTs due to size limit up to %s in in span %s", sz(batchSize), len(toAdd), file.BackupFileEntrySpan.Key, entry.Span)
-					if err := flush(file.BackupFileEntrySpan.Key); err != nil {
+					log.Infof(ctx, "flushing %s batch of %d SSTs due to size limit up to %s in in span (old keyspace) %s", sz(batchSize), len(toAdd), file.BackupFileEntrySpan.Key, entry.Span)
```
Comment (dt) [inline pkg/ccl/backupccl/restore_online.go:231]: are we missing a new verb here?

also `in in`

File: pkg/ccl/backupccl/restore_planning.go [hunk 4]
```diff
@@ -1877,10 +1875,6 @@ func doRestorePlan(
 		if err := checkManifestsForOnlineCompat(ctx, mainBackupManifests); err != nil {
 			return err
 		}
-		currentClusterVersion := p.ExecCfg().Settings.Version.ActiveVersion(ctx).Version
```
Comment (dt) [inline pkg/ccl/backupccl/restore_planning.go:1880]: why did this get cut?

**Review by dt — APPROVED** [review]

File: pkg/ccl/backupccl/restore_planning.go [hunk 5]
```diff
@@ -1362,10 +1362,8 @@ func restorePlanHook(
 		}
 	}
 
-	if restoreStmt.Options.ExperimentalOnline && !restoreStmt.Targets.TenantID.IsSet() {
-		// TODO(ssd): Disable this once it is less annoying to
-		// disable it in tests.
-		log.Warningf(ctx, "running non-tenant online RESTORE; this is dangerous and will only work if you know exactly what you are doing")
+	if restoreStmt.Options.ExperimentalOnline && (restoreStmt.Targets.TenantID.IsSet() || restoreStmt.Options.IncludeAllSecondaryTenants != nil) {
```
Comment (dt) [inline pkg/ccl/backupccl/restore_planning.go:1365]: I think we should add a GA blocker to fix this "bug"

File: pkg/ccl/backupccl/restore_online.go (hunk 3 above)
Comment (msbutler) [reply pkg/ccl/backupccl/restore_online.go:231]: done.

File: pkg/ccl/backupccl/restore_planning.go (hunk 4 above)
Comment (msbutler) [reply pkg/ccl/backupccl/restore_planning.go:1880]: great question. added it back.

File: pkg/ccl/backupccl/restore_planning.go (hunk 5 above)
Comment (msbutler) [reply pkg/ccl/backupccl/restore_planning.go:1365]: done https://github.com/cockroachdb/cockroach/issues/120460

**Comment by msbutler** [comment]
TFTR!

bors r=dt

**Pull request merged**